src/Units/UnitContext.java

```diff
@@ -35,15 +38,26 @@
   public int terrainStars = 0;
 
   public WeaponModel weapon;
+  public int rangeMin = -1, rangeMax = -1;
 
-  public final List<UnitModifier> mods = new ArrayList<UnitModifier>();
+  public final List<UnitModifier> mods = new ArrayList<>();
 
-  public UnitContext(GameMap map, Unit u, WeaponModel w, int x, int y)
+  public UnitContext(GameMap map, Unit u)
+  {
+    this(map, u, null);
+  }
+  public UnitContext(GameMap map, Unit u, WeaponModel w)
+  {
+    this(map, u, w, null, new XYCoord(u));
+  }
+  public UnitContext(GameMap map, Unit u, WeaponModel w, GamePath path, XYCoord coord)
```

Comment:
Why have a constructor with both a path and a coord instead of just using the path's end?

Reply:
The `path` argument can be null.
This is used, for instance, while you have your Rockets move area up, and pan to somewhere else to consider what area you could cover.

Reply:
Hm. Then I'd probably have a constructor with a coord and no path, just to avoid needing dummy parameters.

Can we pull in named parameters from Python? That's a cool feature I'd like to just have everywhere.

Reply:
It's mostly pointless to do that - the only instance where `null` is passed in explicitly is in `AICombatUtils.findThreatPower()` for the case where the unit can't move-and-fire.
That case could easily pass in a path as well. I was just lazy.

> named parameters

Scala has them. It runs on the JVM and is interoperable with Java: https://www.geeksforgeeks.org/scala-named-arguments/

Reply:
> It's mostly pointless to do that

It would help with clarity - it's confusing to have a constructor that demands two apparently-conflicting parameters with no guidance on when one should be provided over the other. It would be cleaner--or at least clearer at point of use-- to just have two constructors to accept either a path or the endpoint.

> That case could easily pass in a path as well. I was just lazy.

It's not lazy to not do useless work. It's pointless at best, and worse if it creates confusion and more unnecessary work.

Reply:
> It would help with clarity - it's confusing to have a constructor that demands two apparently-conflicting parameters with no guidance on when one should be provided over the other.

Either can be `null`, and that is how you indicate how much information you're interested in providing to whatever calculation you're doing.
> It would be cleaner--or at least clearer at point of use-- to just have two constructors to accept either a path or the endpoint.

I dunno about that. The callsites currently (mostly) just pass in the endpoint and `Utils.findShortestPath()` for that endpoint - the caller doesn't necessarily know or care whether `path` is null.
...actually, that's a solid point. If I'm just calling `Utils.findShortestPath()` with arguments I already have, that's overload material.

> It's pointless at best, and worse if it creates confusion and more unnecessary work.

It probably should be there, per my comments in the conversation where you said YAGNI.

Reply:
Are you still making a change for this? Wouldn't hurt to beef up the docs for the six constructors while you're at it.

Reply:
I was originally planning to have the coord-only overload call `Utils.findShortestPath()`, but you pointed out in the other thread that that's a terrible idea.
I should add docs on the constructors, though.